### PR TITLE
tgui: DNA Console Overhaul Project

### PR DESCRIPTION
**Pull request by Timberpoes**
## About The Pull Request

DNA Console UI overhaul project. Full conversion to tgui. Almost complete rewrite of DNA Console codebase. Done with LawyerCode (TM). May require extensive oversight to conform to necessary standards.

Functionality is essentially the same, although there are various little QoL features and changes. In particular: Disks now have a lot more functionality inside the DNA Console and can now be used to combine mutations (which will save to the Disk). Disks also no longer rename themselves when Genetic UI/UE data was stored on them, and this data can be viewed.

Streamlined overall use of DNA Consoles. More information is accessible from more places.

rDWLtQ32yN
WGMR9xhh0X

Also touches DNA Scanners and adds additional quality-of-life functionality. Previously the Scanner would look across all cardinal directions and perform an action on the first Console it found. DNA Scanners and Consoles are now actively linked rather than using the old method. This lays the groundwork for having multiple DNA Scanners attached to a single DNA Console, if you'd ever want that for any reason.

@stylemistake Also added additional tgui Dropdown functionality for hiding the scroll bar and chevrons. Honestly not sure I've done it in a robust way. I'm happy to collaborate further now my PR is live and you can see any code changes I've made to tgui-related stuff.

## Why It's Good For The Game

Conversion to tgui. More flexible interface. Foundation for future Genetics changes.

## Todo List and Other Issues
ISSUE: The monkey activator did not transfer the Reinforced chromosome correctly when recording the test gif. Need to look into this - Is this universal between all activators or is this unique to monkey ones?
UPDATE: Only monkey injectors do not transfer chromosomes. Look into this a bit further.
UPDATE: code/modules/mob/transform_procs.dm - monkeyize() and humanize() do not transfer extra mutations, chromosomes or any other mutation metadata beyond the genetic sequence state of the mob's default mutations. Fix could include full mutation transfer or activating the monkey gene could effectively wipe all the player's mutations clean as it does now, but could especially retain the genome attached to the Monkified mutation. This would mean you could now reinforce monkified to make it immune to mutadone?

TODO: Full suite of DMDOC references for DNA Console code.
TODO: Extensive testing. THIS INCLUDES MULTIPLE PEOPLE TRYING TO USE THE SAME CONSOLE AT THE SAME TIME. Guards have been put in code to attempt to prevent any fuckery. Make sure they are adequate.
TODO: Scour codebase for testing and debug code. Remove this code if any remains before merging.

BONUS: Persistant per-user options including colour configuration/colourblind settings.

## Changelog
:cl:
tweak: DNA Consoles have received hardware upgrades and have a shiny new interface as a result.
tweak: New hardware has dropped legacy support for reading the genetic sequences of the dead and has only limited support for reading the genetic sequences of monkeys.
/:cl:

### CODE REVIEW

**Comment by Iamgoofball** [comment]
just put in joker from persona 5's face for the joker button op

File: code/game/machinery/computer/dna_console.dm [hunk 1]
```diff
@@ -100,850 +129,1050 @@
...
+
+			if(!(adv_inj in injector_selection))
+				return
+
+			var/bref = params["mutref"]
+
+			var/datum/mutation/human/HM = get_mut_by_ref(bref)
+
+			if(HM)
+				injector_selection[adv_inj].Remove(HM)
```
Comment (spookydonut) [inline code/game/machinery/computer/dna_console.dm]: you need to change `var/list/injector_selection` to `var/list/list/injector_selection`

**Review by spookydonut — CHANGES_REQUESTED** [review]
also fix these
loading RandomZLevels/research.dmm
_maps/RandomZLevels/research.dmm:1238:error: invalid variable name: fields
_maps/RandomZLevels/research.dmm:1614:error: invalid variable name: fields

File: code/game/machinery/computer/dna_console.dm (hunk 1 above)
Comment (Timberpoes) [reply code/game/machinery/computer/dna_console.dm]: Change has been made.

**Comment by Timberpoes** [comment]
Updated research.dmm based on my changes to var names in genetics disks.

@Time-Green _maps\RandomZLevels\research.dmm - Two Genetics data disks at 1238 and 1614. I've removed displaying labels in favour of just displaying the associated names. However, do you recall what the SE is? These disks were added in #18199 back in 2016, so they could be holdouts from the old-old genetics?

**Comment by Time-Green** [comment]
> Updated research.dmm based on my changes to var names in genetics disks.
> 
> @Time-Green _maps\RandomZLevels\research.dmm - Two Genetics data disks at 1238 and 1614. I've removed displaying labels in favour of just displaying the associated names. However, do you recall what the SE is? These disks were added in #18199 back in 2016, so they could be holdouts from the old-old genetics?

They were leftovers from old genetics. The SE was how old mutations were stored, it’s probably been broken for ages

**Comment by Timberpoes** [comment]
> 
> 
> > Updated research.dmm based on my changes to var names in genetics disks.
> > _maps\RandomZLevels\research.dmm - Two Genetics data disks at 1238 and 1614. I've removed displaying labels in favour of just displaying the associated names. However, do you recall what the SE is? These disks were added in #18199 back in 2016, so they could be holdouts from the old-old genetics?
> 
> They were leftovers from old genetics. The SE was how old mutations were stored, it’s probably been broken for ages

Thanks. Not a high priority but I added a couple of TODOs to test them.

**Comment by trollbreeder** [comment]
i fucking love ya mate and i hope this gets finished and merged

**Comment by emptyexpression** [comment]
Looks really nice! My only critique is that some of the colors you use aren't ideal from a readability perspective. Especially the A-T pairs being white text on a bright yellow. 

Here's a quick edit with the pair backgrounds darkened:
colors_after

**Comment by stylemistake** [comment]
This interface could use a lot of work in UX department, moving stuff around to make them easier to read and use.

**Comment by stylemistake** [comment]
Yo, here's a sketch of what I think is a better layout.

image

Storage-related buttons obviously go to storage, maybe you can plot Eject Disk button somewhere more suitable, this pic is just for guidance.

**Comment by Timberpoes** [comment]
> 
> 
> Looks really nice! My only critique is that some of the colors you use aren't ideal from a readability perspective. Especially the A-T pairs being white text on a bright yellow.
> 

Agreed. The colours are just a proof-of-concept.

image

This is the colour palette I have access to. What would you suggest utilising from this for AT/GC pairs? Here's an example with "good" and "bad" for the DNA Icon outlines and "green" instead of "olive" for AT pairs.

image

I do actually intend to make this customisable as a strech goal. There's no reason any of these values need to be hardcoded in the final iteration. Colours are arbitrary and can be assigned procedurally.

image 
image



> 
> 
> This interface could use a lot of work in UX department, moving stuff around to make them easier to read and use.

I am not a UI/UX person and I'm not a professional coder. If you have any design ideas, throw them my way along with a rough sketch of how you'd go about implementing them. (e.g. - "Use a Flex with Sections of LabeledLists")

I wouldn't even begin to pretend to understand how to represent the large amounts of information that DNA Consoles have to handle in an effective manner. I don't know how to approach making the UI scalable when expanding things.

For example - What if we increased the mutation storage to higher than 6? What if upgrades could increase mutation storage capacity? What if it was made infinite? My current setup wouldn't be able to handle that.

The same applies to... Most every other thing in DNA Consoles. What if the Chromosome storage limit was increased? What if the default number of mutations on a human was increased from 8? You get the idea.

I wanted to use this total conversion as a springboard to allow more changes to Genetics and some of the above scenarios could play into it. I figured re-arranging the elements would be simple enough once I got the DNA Console code laid out.

**Comment by emptyexpression** [comment]
Ah, I didn't know tgui was restricted to those specific colors. I think it looks fine with the green you chose though.

> HELPME: AT/GC/X/JOKER - What is a good way to implement this? Other ideas I've floated are Buttons that cycle between the letters (How would I do JOKER?) and Input Buttons where people can just type in the letter.

After reading this earlier I wanted to see if it'd be possible to rework the core genetics mechanic with just a button UI, and I came up with this simple game/idea:
https://jsfiddle.net/b9y8zqge/

I don't think this is better than the existing system as is now, but I think there's some potential to use the concept of rotating or shifting DNA blocks as part of the interface.

**Comment by Timberpoes** [comment]
> 
> 
> Ah, I didn't know tgui was restricted to those specific colors. I think it looks fine with the green you chose though.
> 
> > HELPME: AT/GC/X/JOKER - What is a good way to implement this? Other ideas I've floated are Buttons that cycle between the letters (How would I do JOKER?) and Input Buttons where people can just type in the letter.
> 
> After reading this earlier I wanted to see if it'd be possible to rework the core genetics mechanic with just a button UI, and I came up with this simple game/idea:
> https://jsfiddle.net/b9y8zqge/
> 
> I don't think this is better than the existing system as is now, but I think there's some potential to use the concept of rotating or shifting DNA blocks as part of the interface.

It's definitely implementable in tgui - However your mockup is missing something from a gameplay perspective.

Instead of adding a tickbox for an AT or GC pair, can you implement a random genetic sequence that has to be solved? You only get a tick box if you get the AT and/or GC correctly paired, in the correct order and in the correct column (as with the current system) - I believe through playtesting that, you'll probably find this little minigame may turn into a frustrating clickfest. I love it as an idea though.

File: code/game/machinery/computer/dna_console.dm [hunk 2]
```diff
@@ -100,850 +129,1050 @@
...
+				instability_total += I.instability * GET_MUTATION_STABILIZER(I)
+
+			if(instability_total > max_injector_instability)
+				to_chat(usr,"<span class='warning'>Advanced injector mutations too instable.</span>")
+				return
+
+			var/datum/mutation/human/A = new HM.type()
+			A.copy_mutation(HM)
+			injector_selection[adv_inj] += A
+			to_chat(usr,"<span class='notice'>Mutation succesfully added to advanced injector.</span>")
```
Comment (bobbahbrown) [inline code/game/machinery/computer/dna_console.dm]: successfully

Comment (Timberpoes) [reply code/game/machinery/computer/dna_console.dm]: Found all instances of succes and have successfully essed them up. Will be updating as part of larger code documentation I'm currently doing.

**Comment by Timberpoes** [comment]
Streamlining process continues.

image

A lot of extra clutter has been removed. Transitioning from a tabby to a buttony interface with persistent state to come soon after.

New addition as requested by @kriskog - When the Monkified mutation is active, the DNA Console will no longer display the gene sequences of any other mutations. This is a simple "If they are a monkey, display the text instead of the gene sequence" check. Did the same for dead people too.

image

jZ04IxThQp

image

**Comment by Timberpoes** [comment]
Updated interface concept. Cleaner. More streamlined. I'm going to apply this design pattern to the rest of the console.

xxE8oNOERb

**Comment by stylemistake** [comment]
- Partially refactored the interface (top-level layout, storage and sequencer).
- Implemented a gene cycle component.
  - **Click:** next gene
  - **Right Click:** previous gene
  - **Ctrl+Click:** place X
  - **Shift+Click:** place a question mark
  - **Joker, then Click:** reveal a gene
- Reworked tgui view state, which is now much simpler and much more generic, net decrease in LoC.

**Comment by Timberpoes** [comment]
Removed list() versions of strings DM-side.
Removed used TGUI-exclusive constants from data object DM-side. Currently unused ones still remain, as a reminder that they're probably part of some forgotten logic in the DNA Console.
Fixed other DM data structures to serialise as JS Arrays instead of Objects. Removed toArray() code in JS and added guards to prevent null or undefined arrays as a result.
Minor TGUI logic fixes for some edge cases including TGUI guards to prevent various actions being performed on mutations that are discovered but inactive.
Added some comments in DnaConsole.js outlining some minor notes and TODOs.

Weird "features" to be aware of in the current refactor -> Gene sequencer and mutation storage share mutation code.

Problem is that the actions available on stored mutations are contextually different to those of mutations in the sequencer:-

Stored mutations cannot have chromosomes applied to them, that is unique to sequencer mutations.
Stored mutations cannot be added to advanced injectors, that is unique to sequencer mutations.

Changing these two features to be applicable to stored mutations would constitute a balance change (a minor buff to Genetics and also being an overall usability improvement).

Likewise the above has added extraneous buttons to the interface. Console-stored mutations have a permanently greyed out "Save to Console" button, which doesn't make sense. Disk-stored mutations have the same but with a "Save to Disk" button.

Because only active mutations can even BE stored to console, I'm going to do another commit shortly to add some additional data to console/disk stored mutations defining them as Active = TRUE to restore the ability to copy mutations from Disk to Console and vice-versa.

**Comment by stylemistake** [comment]
More or less finished the rework. Few bugs remain, but it's quite usable now!

[Video demo](https://cdn.discordapp.com/attachments/326831214667235328/696082068945240075/2020-04-04_22-34-40.mp4) (skip to 2:16 for the actual console footage)

**Comment by Timberpoes** [comment]
I'll be giving this a full suite of testing on my server to make sure everything performs as expected. I'll drop a reply here when said testing is complete.

I'm also going to try hunt down some geneticists for feedback too, but holy GitHub friendly expletives this new system feels amazingly slick to use.

Full thanks to all parties involved in this project and the patience everyone has shown while guiding me.

**Comment by PKPenguin321** [comment]
I'd love if you could add stored mutations to advanced injectors, if possible

**Comment by Timberpoes** [comment]
> I'd love if you could add stored mutations to advanced injectors, if possible

It's possible (infact, I coded specifically to add it in a future PR)

image

I can expand that to include disk and console checks too and it would instantly enable this functionality. I wanted to keep this as close as possible to the original HTML interface functionality, however I have taken some creative liberties already - For example disks will be more flexible as I plan to allow mutation combining to pool from both the disk and DNA Console storage.

Any pseudo-balance changes like that (making previously annoying and fiddly things more accessible or have additional code paths enabling the annoying and fiddly thing) I'm going to play by the book and seek appropriate maintainer approval to include in this PR.

Or if you can hunt down someone who can okay any minor pseudo-balance changes like that for me, I'll document any such changes for final approval before merging.

**Comment by Timberpoes** [comment]
I think this FINALLY may be in a pseudo-complete state.

## Balance changes:

- Mutations can be combined on both disk and console. The combining pool pulls from both. You can combine disk mutations with console mutations and save on either medium depending on which mutation is your source. Geneticists no longer need to juggle mutations so hard between disk and console.

- Unsolved genes ('X') from when the mutation was first initialised are highlighted. This is persistent until the mutation is solved. Genetecists who take a break or get dragged away while brute-forcing XX pairs will now have an easier time returning to the mutation and continuing their brute-force.

**Pull request merged**